### PR TITLE
Container content management commands

### PR DESCRIPTION
**Pull request by gerrod3**
Mega WIP

### CODE REVIEW

File: pulpcore/cli/common/generic.py [hunk 1]
```diff
@@ -326,7 +326,7 @@ def _callback(
 
 
 # based on https://stackoverflow.com/a/42865957/2002471
-units = {"B": 1, "KB": 10**3, "MB": 10**6, "GB": 10**9, "TB": 10**12}
+units = {"B": 1, "KB": 10 ** 3, "MB": 10 ** 6, "GB": 10 ** 9, "TB": 10 ** 12}
```
Comment (mdellweg) [inline pulpcore/cli/common/generic.py:329]: Did black change their mind again, or is your version outdated?

File: pulpcore/cli/common/generic.py [hunk 2]
```diff
@@ -1002,29 +1000,30 @@ def version_callback(
         )
         return repo_ver_ctx
 
-    @click.command("list")
-    @click.option("-t", "--type", "type", type=click.Choice(names), default=names[0])
+    @pulp_command("list")
+    @click.option("--all-types", is_flag=True)
```
Comment (gerrod3) [inline pulpcore/cli/common/generic.py:1004]: I've changed the default behavior of this command group. It might make more sense to pull this out into a separate PR.

File: pulpcore/cli/container/repository.py [hunk 3]
```diff
@@ -44,22 +51,62 @@
...
+        # Check that the version makes sense
+        latest_version = int(repo_ctx.entity["latest_version_href"].split("/")[-2])
+        if not (0 < int(version) <= latest_version):
+            raise click.ClickException(
+                _(
+                    "Please specify a version that is less then the latest version {} and greater"
+                    " than 0"
+                ).format(latest_version)
+            )
+        repo_ctx.meta["version"] = version
```
Comment (gerrod3) [inline pulpcore/cli/container/repository.py:99]: Can you just store what you want in meta? Because that's what I doing.

File: pulpcore/cli/common/generic.py (hunk 1 above)
Comment (mdellweg) [reply pulpcore/cli/common/generic.py]: Well, black said, they are now stable and won't change incompatible anymore...

File: pulpcore/cli/container/repository.py (hunk 3 above)
Comment (mdellweg) [reply pulpcore/cli/container/repository.py:99]: Hmm good question, i don't even see it being used somewhere else.

Comment (mdellweg) [reply pulpcore/cli/container/repository.py:99]: One concern i have with such things is the lifetime, once we get around to more pulp-cli-shell operations, where such a meta object could leak from one command into the next. But we are not there...

Comment (gerrod3) [reply pulpcore/cli/container/repository.py:99]: I can pop the value from the meta object so that it goes back to its original state.

File: pulpcore/cli/ansible/repository.py [hunk 4]
```diff
@@ -121,13 +120,6 @@ def repository(ctx: click.Context, pulp_ctx: PulpContext, repo_type: str) -> Non
...
-    type_option(
-        choices={
-            "collection-version": PulpAnsibleCollectionVersionContext,
-            "role": PulpAnsibleRoleContext,
-        },
-        default="collection-version",
-    ),
     href_option,
 ]
 content_json_callback = create_content_json_callback(schema=CONTENT_LIST_SCHEMA)
```
Comment (mdellweg) [inline pulpcore/cli/ansible/repository.py:125]: I am a bit lost here. This schema is supposed to handle both collections and roles?

File: pulpcore/cli/ansible/repository.py [hunk 5]
```diff
@@ -150,13 +142,6 @@ def repository(ctx: click.Context, pulp_ctx: PulpContext, repo_type: str) -> Non
     The argument prefixed with the '@' can be the path to a JSON file with a list of objects."""
         ),
     ),
-    type_option(
-        choices={
-            "collection-version": PulpAnsibleCollectionVersionContext,
-            "role": PulpAnsibleRoleContext,
-        },
-        default="collection-version",
-    ),
```
Comment (mdellweg) [inline pulpcore/cli/ansible/repository.py:159]: So these commands are supposed to handle all types of content in unison now?

File: pulpcore/cli/common/generic.py (hunk 2 above)
Comment (mdellweg) [reply pulpcore/cli/common/generic.py:1004]: What motivates the change?

File: pulpcore/cli/common/generic.py [hunk 6]
```diff
@@ -1002,29 +1000,30 @@ def version_callback(
...
-        type: Optional[str],
+        all_types: Optional[bool],
         **params: Any,
     ) -> None:
         parameters = {k: v for k, v in params.items() if v is not None}
         parameters.update({"repository_version": version.pulp_href})
-        result = content_contexts[type](pulp_ctx).list(
-            limit=limit, offset=offset, parameters=parameters
-        )
+        ctx_obj = PulpContentContext(pulp_ctx) if all_types else ctx.obj
```
Comment (mdellweg) [inline pulpcore/cli/common/generic.py:1022]: So this will either result in an object of type `PulpContentContext` or `PulpRepositroy`? That sounds wrong to me.
ALso what is the desired behaviour if `all_types` is `False`?

File: pulpcore/cli/ansible/repository.py (hunk 4 above)
Comment (gerrod3) [reply pulpcore/cli/ansible/repository.py:125]: Yes, since collection and roles have the same identifiers (namespace, name, version).

File: pulpcore/cli/ansible/repository.py (hunk 5 above)
Comment (gerrod3) [reply pulpcore/cli/ansible/repository.py:159]: This option is now apart of the content command group. I've moved it back one spot in the command.

File: pulpcore/cli/common/generic.py (hunk 6 above)
Comment (gerrod3) [reply pulpcore/cli/common/generic.py:1022]: No it will always be a subtype of `PulpContentContext`, I should probably use `pass_content_context` since this group's context object will always be set to a content context due to the type option.

File: pulpcore/cli/container/repository.py (hunk 3 above)
Comment (mdellweg) [reply pulpcore/cli/container/repository.py:99]: Shouldn't we return a RepoVersionContext from this parameter?

Comment (gerrod3) [reply pulpcore/cli/container/repository.py:99]: I did want to, but `meta`'s type definition is `Dict[str, str]`.

Comment (mdellweg) [reply pulpcore/cli/container/repository.py:99]: Sorry, i meant the whole callback.

Comment (gerrod3) [reply pulpcore/cli/container/repository.py:100]: Since the API takes either repository or repository version I decided attaching the version to a `RepositoryContext` was easiest way to determine which the user was specifying.

Comment (mdellweg) [reply pulpcore/cli/container/repository.py:100]: Is specifying no version different from specifying the latest version?

Comment (gerrod3) [reply pulpcore/cli/container/repository.py:100]: Looking at the serializer, no they are the same (except you can't specify a repository with 0 versions): https://github.com/pulp/pulp_container/blob/main/pulp_container/app/serializers.py#L527

Comment (gerrod3) [reply pulpcore/cli/container/repository.py]: Updated to return `PulpContainerRepositoryVersionContext`.

File: pulpcore/cli/container/context.py [hunk 7]
```diff
@@ -110,6 +110,39 @@ class PulpContainerRepositoryContext(PulpContainerBaseRepositoryContext):
...
+        add_content: Optional[List[str]] = None,
+        remove_content: Optional[List[str]] = None,
+        base_version: Optional[str] = None,
+    ) -> Any:
+        if remove_content:
+            self.call(
+                "remove", parameters={self.HREF: href}, body={"content_units": remove_content}
+            )
+        if add_content:
+            self.call("add", parameters={self.HREF: href}, body={"content_units": add_content})
```
[the remark this answers is not in the document]
Comment (gerrod3) [reply pulpcore/cli/container/context.py:125]: To which class, the `PulpContainerBaseRepositoryContext`? I don't think so since only the regular container repository can call the add and remove endpoints. Also this modify call is needed since that is what the content factory uses to create the add and remove commands.

File: pulpcore/cli/container/repository.py [hunk 8]
```diff
@@ -45,22 +53,66 @@
...
+    pulp_ctx = ctx.find_object(PulpContext)
+    assert pulp_ctx is not None
+    if len(value) == 0:
+        raise click.ClickException(_("Please pass in a non empty source repository."))
+
+    version: Optional[str] = None
+    pulp_href: Optional[str] = None
+    entity: Optional[EntityDefinition] = None
+    if value.startswith("/"):
+        pattern = rf"^{pulp_ctx.api_path}{PulpContainerRepositoryContext.HREF_PATTERN}"
```
[the remark this answers is not in the document]
Comment (gerrod3) [reply pulpcore/cli/container/repository.py:75]: Technically, but it shouldn't, I'll fix that.

File: pulpcore/cli/container/repository.py [hunk 9]
```diff
@@ -45,22 +53,66 @@
...
+                _("'{value}' is not a valid href for {option_name}.").format(
+                    value=value, option_name=param.name
+                )
+            )
+        pulp_href = value
+        if "versions" in value:
+            pulp_href = value.partition("versions")[0]
+            version = value.split("/")[-2]
+    else:
+        split_value = value.split(":", maxsplit=1)
```
[the remark this answers is not in the document]
Comment (gerrod3) [reply pulpcore/cli/container/repository.py:87]: I'm not sure, the serializer for this operation only allows container repositories to be specified, but doesn't have the same limit on the repository version field. So you couldn't supply a push repository, but you could supply one of its versions. I feel like that has to be an oversight.

File: pulpcore/cli/container/context.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/cli/container/context.py:125]: I was really thinking about `PulpRepositoryContext` now that you say that push-repos cannot modify content, this would add the need for a capability. Maybe not worth the trouble now.

File: pulpcore/cli/container/repository.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/cli/container/repository.py:87]: Yes, i think you must be able to push content to a push repo and subsequently copy it to a "normal" one.

[the remark this answers is not in the document]
Comment (gerrod3) [reply pulpcore/cli/container/repository.py]: Ok, added support for copying from push repositories.

File: pulpcore/cli/container/repository.py [hunk 10]
```diff
@@ -165,3 +247,77 @@ def add_tag(
...
+
+
+@repository.command(allowed_with_contexts=container_context)
+@name_option
+@href_option
+@click.option(
+    "--source",
+    help=_(
+        "Source repository to copy manifests from. Specify in format of href or [<resource_type>:]"
+        "<name>[:<version>]"
```
Comment (mdellweg) [inline pulpcore/cli/container/repository.py:286]: :dragon: optional parts in two places sounds real dangerous. Also i had expected that wherever we have a resource parameter, we'd allow for `[[<plugin>:]<type>:]<name>`.
Also we need to make sure we still allow ":" in the name of the repository with `::reponame:with:colons`.
Is it very unintuitive to use a separate separator for the version?

**Review by mdellweg — APPROVED** [review]
I think, this is it.
